### ISSUE TITLE
Provide support and recommendations for compressed OWL

### ISSUE THREAD

**Issue by cmungall** [OBOFoundry/OBOFoundry.github.io#474]
Some ontologies are becoming v large. E.g. v155 of CHEBI:
http://purl.obolibrary.org/obo/chebi.owl

is ~840M. Other large ontologies are pr (cc @nataled), ncbitaxon.

The large size is causing many pipelines to fail (see for example: https://github.com/geneontology/noctua/issues/486). In the case of CHEBI, this may be compounded by various factors:

 * The ontology file is being served from FTP: https://github.com/OBOFoundry/purl.obolibrary.org/blob/master/config/chebi.yml#L7
    * it would be better if CHEBI were served from HTTP. See for example [this stackoverflow q](https://stackoverflow.com/questions/717200/comparing-http-and-ftp-for-transferring-files)
 * connecting to EBI may be slower at the w/e???
 * The OWLAPI may have problems with remote URLs with large files, possibly compounded if over FTP??

Note that CHEBI provides a gzipped OWL, and we register this: https://github.com/OBOFoundry/purl.obolibrary.org/blob/f03d247b91f90e88ad293872eba406d5c5cfc70f/config/chebi.yml#L9

The OWLAPI handles gzipped owl just fine. I can feed this PURL into the OWLAPI:

http://purl.obolibrary.org/obo/chebi.owl.gz

And it works fine and is *much* faster than the canonical PURL.

Unfortunately, using a non-canonical PURL violates expectations and has to be explicitly wired in to pipelines. We also have no OBO policy on provision of gzipped files at the top level. So far AFAIK CHEBI is the only one to provide this.

Ideally things would just work and client and server could negotiate and send gzipped if this is supported, but not clear how to work this into OBO infrastructure, with a mix of centrally served files (bbop S3) and ontology-specific (e.g. FTP for CHEBI, github for some, etc)

**Comment by kltm** [comment]
Well, IIRC, from the last time this was discussed (and a quick look to refresh my memory), I believe we basically came up with: many web servers (apache, etc.) let you do things like negotiated compression and remapping on the fly pretty easily, but S3 (and CloudFront) pretty much leaves you in a lurch. I think that practically it might be easiest to either have the OWLAPI try negotiation (or just straight attempting to grab various zipped URLs) on it's own (either following a convention or adding more metadata).

**Comment by cmungall** [comment]
One approach that doesn't involve any OWLAPI changes is to supply an OBO-wide catalog file with entries 

```
http://purl.obolibrary.org/obo/chebi.owl => http://purl.obolibrary.org/obo/chebi.owl.gz
```

**Comment by kltm** [comment]
That's a nice idea; I wonder how that could be composed with the GO catalogs that we'll have to produce during the pipeline runs.

**Comment by cmungall** [comment]
We'd compose them such that our own take precedence over general.

I think owltools allows composition - would have to check

but better to compose ahead of time with some generic catalog 
composition tool (To be written)

**Comment by nlharris** [comment]
is this done?

**Comment by alanruttenberg** [comment]
Could this not be accomplished with content negotiation? For instance a
request for http://purl.obolibrary.org/obo/chebi.owl with an Accept header
<https://webmasters.stackexchange.com/questions/31212/difference-between-the-accept-and-content-type-http-headers>
application/gzip would provide the gzipped version. This puts the onus for
getting the gzipped version on the application that is accessing it,
assuming we configure the server to be able to respond with the gzipped
version. There's also the Accept-Encoding
<https://stackoverflow.com/questions/46282063/accept-encodinggzip-and-content-encodinggzip>
header that could be used.

Alan

**Comment by cmungall** [comment]
Hi @alanruttenberg. How would this work for the current ways we have of serving files up across different ontologies

 - S3
 - raw github URLs
 - github downloads
 - FTP (chebi)
 - others?

@kltm said:
> many web servers (apache, etc.) let you do things like negotiated compression and remapping on the fly pretty easily, but S3 (and CloudFront) pretty much leaves you in a lurch

Would this solution require that ontology providers run their own Apache/Nginx server? Many groups don't have resources to do this, and it is an extra point of potential failure.

We could do something clever on the purl server but I think this would be bad for a number of reasons.

I don't think we can close this until we have recommendations (even if the recommendation is: tough, just download the large uncompressed file)

**Comment by alanruttenberg** [comment]
Maybe this is something that can be handled by the PURL server. Attend to
the Accept-Encoding header and add ".gz" as it forwards the request, if
appropriate. The configuration files could have a property that indicates
whether a particular ontology will or won't serve compressed versions.

Alan

**Comment by cmungall** [comment]
This could work. It adds a bit more complexity. But it doesn't require our
server doing any compression/uncompression

I worry about debugging headaches further down the line though...

**Comment by nataled** [comment]
PRO provides that very solution: appending '.gz' to the owl PURL will grab the compressed version, which we supply on our end. I believe @matentzn uses this (IIRC, it was provided at his request).

**Comment by alanruttenberg** [comment]
On Tue, Dec 1, 2020 at 2:36 AM Chris Mungall <notifications@github.com>
wrote:

> This could work. It adds a bit more complexity. But it doesn't require our
> server doing any compression/uncompression
>
> I worry about debugging headaches further down the line though...
>

What could go wrong with change that can't go wrong now? I guess James
would have a better sense. It doesn't seem, on the face of it, to be
particularly complicated.
Probably could have a switch that turned off the behavior if there's an
issue. Clients would get uncompressed versions for a while until it is
fixed.

**Comment by jamesaoverton** [comment]
The PURL system is design to be as stupid and transparent as possible. I don't want to start inspecting headers to determine redirects. It's not that it would be hard, but it would be one more layer of complexity. As simple as the system is right now, it still confuses people.

My preference would be for OBO to host a compressed version of the primary OWL file for each project on a reliable host such as S3 at a new, standard location such as <http://purl.obolibrary.org/obo/foo.owl.gz> (or maybe <http://purl.obolibrary.org/mirror/foo.owl.gz>). That approach has its own complexity costs, but would solve a wider range of problems with slow or broken downloads. See also #494 and https://github.com/OBOFoundry/purl.obolibrary.org/issues/162

**Comment by alanruttenberg** [comment]
I agree that having it all served from one place would be optimal. There
are costs associated so one question is who bears the cost?
I do think that the choice of compressed or not should be decided by Accept
headers, so if it's going to be S3 we want to make sure it can do that.

Alan
